### PR TITLE
Add compute_machine_types datasource

### PR DESCRIPTION
**Pull request by rifelpet**
This implements a plural version of https://github.com/hashicorp/terraform-provider-google/issues/5606 which should solve the same use-cases while also being more flexible.

Specifically, a plural datasource allows a terraform module to be used across regions without needing to override a machine_type variable based on [machine type availability](https://cloud.google.com/compute/docs/regions-zones#available). The plural datasource can return an empty list rather than fail the terraform operation if a specific machine type isn't available in the given zone.

Regarding the test, similar to other read-only APIs I'm only checking the existence of attributes rather than requesting a specific machine type and checking its values, since different test environments may have different machine type availability.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-datasource
`google_compute_machine_types`
```

### CODE REVIEW

**Comment by rifelpet** [comment]
Gentle ping @slevenick, can you take a look at this when you have a chance? Thanks!

File: mmv1/third_party/terraform/website/docs/d/compute_machine_types.html.markdown [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+* `project` (Optional) - Project from which to list available zones. Defaults to project declared in the provider.
+
+## Attributes Reference
+
+The following attribute is exported:
+
+* `machine_types.0.name` - The name of the machine type.
+* `machine_types.0.guest_cpus` - The number of virtual CPUs that are available to the instance.
+* `machine_types.0.memory_mb` - The amount of physical memory available to the instance, defined in MB.
+* `machine_types.0.bundled_local_ssds.0.default_interface` - (Beta only) The default disk interface if the interface is not specified.
```
Comment (slevenick) [inline mmv1/third_party/terraform/website/docs/d/compute_machine_types.html.markdown:102]: Our beta-only fields typically have links here. See https://github.com/hashicorp/terraform-provider-google-beta/blob/main/website/docs/r/artifact_registry_repository.html.markdown?plain=1#L379

File: mmv1/third_party/terraform/website/docs/d/compute_machine_types.html.markdown [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+* `filter` (Required) - A filter expression that filters machine types listed in the response.
+* `zone` (Required) - Zone from which to list machine types.
+* `project` (Optional) - Project from which to list available zones. Defaults to project declared in the provider.
+
+## Attributes Reference
+
+The following attribute is exported:
+
+* `machine_types.0.name` - The name of the machine type.
```
Comment (slevenick) [inline mmv1/third_party/terraform/website/docs/d/compute_machine_types.html.markdown:99]: We typically document the object structure without the indexes. Can you change this to be more similar to a typical resource/datasource?

**Comment by rifelpet** [comment]
Thanks for the pointers @slevenick, I believe I've addressed your feedback. Please take another look when you get a chance.

File: mmv1/third_party/terraform/services/compute/data_source_google_compute_machine_types.go.erb [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+	}
+	if err := d.Set("zone", zone); err != nil {
+		return diag.FromErr(fmt.Errorf("Error setting zone: %w", err))
+	}
+
+	query := url.Values{}
+	if filter != "" {
+		query.Set("filter", filter)
+	}
+	id := url.URL{
```
[the remark this answers is not in the document]
Comment (rifelpet) [reply mmv1/third_party/terraform/services/compute/data_source_google_compute_machine_types.go.erb]: Makes sense. I've updated this to use your Sprintf suggestion.

**Review by slevenick — APPROVED** [review]
Thanks!

**Pull request merged**